upgrades/upgrade_testing_framework/core/expectation.py

```diff
@@ -0,0 +1,96 @@
+from pathlib import Path
+import json
+from typing import List
+import operator
+
+from upgrade_testing_framework.core.versions_engine import EngineVersion
+
+class KnowledgeBaseDoesntExistException(Exception):
+    def __init__(self, knowledge_base_path):
+        super().__init__(f"The path specified for the knowledge base doesn't exist or is not a directory: {knowledge_base_path}")
+
+class ExpectationCantReadFileException(Exception):
+    def __init__(self, expecation_path, original_exception):
+        super().__init__(f"Unable to read test config file at path {expecation_path}.  Details: {str(original_exception)}")
+
+class ExpectationFileNotJSONException(Exception):
+    def __init__(self, expecation_path, original_exception):
+        super().__init__(f"The test config at path {expecation_path} is not parsible as JSON.  Details: {str(original_exception)}")
+
+class ExpectationMissingIdException(Exception):
+    def __init__(self):
+        super().__init__("An expectation was missing its id") # TODO make this better
+
+
+COMP_OPERATION = {
+    "gt": operator.gt,
+    "lt": operator.lt,
+    "gte": operator.ge,
+    "lte": operator.le
+}
+
+
+class Expectation:
+    def __init__(self, raw_expectation: dict):
+        self.id = raw_expectation.get("id", None)
+        if self.id is None:
+            raise ExpectationMissingIdException()
+        self.description = raw_expectation.get("description", None)
+        self.version_filter = raw_expectation.get("versions", None)
+        
+
+    def is_relevant_to_version(self, version: EngineVersion) -> bool:
+        # If no version_filter, assume relevant to all versions
+        if self.version_filter is None:
+            return True
+
+        version_filters = self.version_filter if type(self.version_filter) is list else [self.version_filter]
+
+        # Version filters have sections which are ORed together, each of which has
+        # components (gt/gte and/or lt/lte), which are ANDed together.
+        for section in version_filters:
```

Comment:
Can we add DEBUG-level log statements to help troubleshoot when we inevitably have a subtly-wrong set of conditionals?

Reply:
Good call. Added in the forthcoming revision, and makes the logic very traceable.

```
2022-12-23 22:29:04.503914 - upgrade_testing_framework.core.framework_step - [SelectExpectations] Loading Knowledge Base and filtering for relevant expectations
2022-12-23 22:29:04.505845 - upgrade_testing_framework.core.expectation - Comparing version ES_7_10_2 to the version filter [{'gte': 'ES_7_0_0', 'lt': 'OS_2_4_0'}].
2022-12-23 22:29:04.505982 - upgrade_testing_framework.core.expectation - Version ES_7_10_2 applies because it satisfies {'gte': 'ES_7_0_0', 'lt': 'OS_2_4_0'}.
2022-12-23 22:29:04.506091 - upgrade_testing_framework.core.expectation - Comparing version ES_7_10_2 to the version filter [{'gte': 'OS_2_4_0'}, {'lt': 'ES_7_0_0'}].
2022-12-23 22:29:04.506210 - upgrade_testing_framework.core.expectation - Version ES_7_10_2 does not satisfy conditon `gte OS_2_4_0`
2022-12-23 22:29:04.506270 - upgrade_testing_framework.core.expectation - Version ES_7_10_2 does not satisfy conditon `lt ES_7_0_0`
2022-12-23 22:29:04.506321 - upgrade_testing_framework.core.expectation - Version ES_7_10_2 does not apply because it failed to satisfy any sections of the version filter.
2022-12-23 22:29:04.506385 - upgrade_testing_framework.core.expectation - Comparing version OS_1_3_6 to the version filter [{'gte': 'OS_2_4_0'}, {'lt': 'ES_7_0_0'}].
2022-12-23 22:29:04.506435 - upgrade_testing_framework.core.expectation - Version OS_1_3_6 does not satisfy conditon `gte OS_2_4_0`
2022-12-23 22:29:04.506482 - upgrade_testing_framework.core.expectation - Version OS_1_3_6 does not satisfy conditon `lt ES_7_0_0`
2022-12-23 22:29:04.506528 - upgrade_testing_framework.core.expectation - Version OS_1_3_6 does not apply because it failed to satisfy any sections of the version filter.
```